File: ws_client.rb
#!/usr/bin/env ruby

require 'rubygems'
require 'rest-client'
require 'json'

# Identification sur le web_service en tant que société
begin
  response = RestClient.post("https://ws.agiloffice.fr/demo/session_firm_login?password=xxxxxx", {:accept => 'application/json'})
  if response.code == 200
    puts "Ok : identification effectuée"
    cookie_str = response.headers[:set_cookie].first.split(';').first

    # Décrochage du poste 8003
    begin
      response = RestClient.post("https://ws.agiloffice.fr/demo/phone_answer?phone=8003", {}, {:accept => 'application/json', :cookie => cookie_str})
      if response.code == 200
        puts "Ok : décrochage du poste effectué : #{response.body}"
      else
        puts "Erreur #{response.code} : poste non décroché : #{response.body}"
      end
    rescue => e
      if e.is_a?(RestClient::Exception)
        puts "Erreur #{e.response.code} : poste non décroché : #{e.response.body}"
      else
        puts "Exception : #{e.message}\n" + e.backtrace.join("\n")
      end
    end

    # Raccrochage du poste 8003
    begin
      response = RestClient.post("https://ws.agiloffice.fr/demo/phone_hangup?phone=8003", {}, {:accept => 'application/json', :cookie => cookie_str})
      if response.code == 200
        puts "Ok : raccrochage du poste effectué : #{response.body}"
      else
        puts "Erreur #{response.code} : poste non raccroché : #{response.body}"
      end
    rescue => e
      if e.is_a?(RestClient::Exception)
        puts "Erreur #{e.response.code} : poste non raccroché : #{e.response.body}"
      else
        puts "Exception : #{e.message}\n" + e.backtrace.join("\n")
      end
    end
  else
    puts "Erreur #{response.code} : identification échouée : #{response.body}"
  end

  # Récupération de l'url du serveur de push
  begin
    response = RestClient.get("https://ws.agiloffice.fr/demo/push_server_url", {:accept => 'application/json', :cookie => cookie_str})
    if response.code == 200
      puts "Ok : url du push_server récupérée : #{response.body}"
    else
      puts "Erreur #{response.code} : url du push_server non récupérée : #{response.body}"
    end
  rescue => e
    if e.is_a?(RestClient::Exception)
      puts "Erreur #{e.response.code} : url du push_server non récupérée : #{e.response.body}"
    else
      puts "Exception : #{e.message}\n" + e.backtrace.join("\n")
    end
  end

  # Envoi d'un fax
  begin
    response = RestClient.post("https://ws.agiloffice.fr/demo/send_fax?from=0123456789&to=9876543210",
      {:file => File.new("#{File.dirname(__FILE__)}/document.pdf", 'rb')},
      {:accept => 'application/json', :cookie => cookie_str})
    if response.code == 200
      puts "Ok : fax envoyé : #{response.body}"
      fax_id = JSON.parse(response.body)['fax_id']
      # Récupération de l'état du fax envoyé
      begin
        response = RestClient.get("https://ws.agiloffice.fr/demo/get_fax_status?fax_id=#{fax_id}", {:accept => 'application/json', :cookie => cookie_str})
        if response.code == 200
          puts "Ok : statut du fax récupéré : #{response.body}"
        else
          puts "Erreur #{response.code} : statut du fax non récupérée : #{response.body}"
        end
      rescue => e
        if e.is_a?(RestClient::Exception)
          puts "Erreur #{e.response.code} : statut du fax non récupérée : #{e.response.body}"
        else
          puts "Exception : #{e.message}\n" + e.backtrace.join("\n")
        end
      end
    else
      puts "Erreur #{response.code} : fax non envoyé : #{response.body}"
    end
  rescue => e
    if e.is_a?(RestClient::Exception)
      puts "Erreur #{e.response.code} : fax non envoyé : #{e.response.body}"
    else
      puts "Exception : #{e.message}\n" + e.backtrace.join("\n")
    end
  end
rescue => e
  if e.is_a?(RestClient::Exception)
    puts "Erreur #{e.response.code} : identification échouée : #{e.response.body}"
  else
    puts "Exception : #{e.message}\n" + e.backtrace.join("\n")
  end
end

File: ws_client.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from httplib2 import Http
from urllib import urlencode
import mimetools, mimetypes
from cStringIO import StringIO
import os, stat
import json

def multipart_encode(fd):
  boundary = mimetools.choose_boundary()
  buf = StringIO()
  file_size = os.fstat(fd.fileno())[stat.ST_SIZE]
  filename = fd.name.split('/')[-1]
  buf.write('--%s\r\n' % boundary)
  buf.write('Content-Disposition: form-data; name="%s"; filename="%s"\r\n' % ('file', filename))
  buf.write('Content-Type: application/pdf\r\n')
  fd.seek(0)
  buf.write('\r\n' + fd.read() + '\r\n')
  buf.write('--' + boundary + '--\r\n\r\n')
  buf = buf.getvalue()
  return boundary, buf

h = Http()

# Identification sur le web_service en tant que société
resp, content = h.request("http://ws.agiloffice.fr/demo/session_firm_login?password=xxxxxx", "POST", headers = {'Accept':'application/json'})
if resp.status == 200:
  print "Ok : identification effectuée"
  cookie_str = resp['set-cookie'].split(';')[0]
  
  # Raccrochage du poste 8003
  resp, content = h.request("http://ws.agiloffice.fr/demo/phone_hangup?phone=8003", "POST", headers = {'Accept':'application/json', 'Cookie':cookie_str})
  if resp.status == 200:
    print "Ok : raccrochage du poste effectué : " + content
  else:
    print "Erreur " + resp['status'] + " : poste non raccroché : " + content
  
  # Récupération de l'url du serveur de push
  resp, content = h.request("http://ws.agiloffice.fr/demo/push_server_url", "GET", headers = {'Accept':'application/json', 'Cookie':cookie_str})
  if resp.status == 200:
    print "Ok : url du push_server récupérée : " + content
  else:
    print "Erreur " + resp['status'] + " : url du push_server non récupérée : " + content
  
  # Envoi d'un fax
  boundary, data = multipart_encode(open('document.pdf'))
  resp, content = h.request("http://ws.agiloffice.fr/demo/send_fax?from=0123456789&to=9876543210", "POST", body = data, \
    headers = {'Accept':'application/json', 'Cookie':cookie_str, 'Content-type':'multipart/form-data; boundary=' + boundary})
  if resp.status == 200:
    print "Ok : fax envoyé : " + content
    fax_id = json.loads(content)['fax_id']

    # Récupération de l'état du fax envoyé
    resp, content = h.request("http://ws.agiloffice.fr/demo/get_fax_status?fax_id=%d" % fax_id, "GET", headers = {'Accept':'application/json', 'Cookie':cookie_str})
    if resp.status == 200:
      print "Ok : statut du fax récupéré : " + content
    else:
      print "Erreur " + resp['status'] + " : statut du fax non récupérée : " + content
  else:
    print "Erreur " + resp['status'] + " : fax non envoyé : " + content
else:
  print "Erreur " + resp['status'] + " : identification échouée : " + content
